examples/resnet-concurrent.cpp

```diff
@@ -0,0 +1,164 @@
+llvm::cl::opt<unsigned> numDevices("numDevices",
+                                   llvm::cl::desc("Number of Devices to use"),
+                                   llvm::cl::init(7), llvm::cl::value_desc("N"),
+                                   llvm::cl::cat(category));
+} // namespace
+
+/// Run ResNet concurrently on a fixed number of CPU Devices
+int main(int argc, char **argv) {
+  llvm::cl::ParseCommandLineOptions(
+      argc, argv, "Run ResNet concurrently on a fixed number of CPU devices");
+
+  llvm::outs() << "Loading files from " << inputDirectory << "\n";
+
+  llvm::outs() << "Initializing " << numDevices << " CPU Devices.\n";
+  std::vector<std::unique_ptr<CPUDeviceManager>> devices;
+  for (int i = 0; i < numDevices; ++i) {
+    devices.emplace_back(std::make_unique<CPUDeviceManager>());
+    devices[i]->init();
+  }
+
+  Module module;
+  Function *F = module.createFunction("resnet50");
+  TypeRef inputType(module.uniqueType(ElemKind::FloatTy, {1, 3, 224, 224}));
+
+  llvm::outs() << "Loading resnet50 model.\n";
+
+  const char inputName[] = "gpu_0/data";
+  Caffe2ModelLoader loader("resnet50/predict_net.pb", "resnet50/init_net.pb",
+                           {inputName}, {inputType}, *F);
+  Placeholder *input =
+      llvm::cast<Placeholder>(cantFail(loader.getNodeValueByName(inputName)));
+  Placeholder *output = cantFail(loader.getSingleOutput());
+  auto *backend = createBackend(BackendKind::CPU);
+
+  llvm::outs() << "Starting compile.\n";
+  backend->optimizeFunction(CompilationMode::Infer, F);
+  auto compiledFunction = backend->compile(F);
+
+  FunctionMapTy functionMap;
+  functionMap.emplace("resnet50", compiledFunction.get());
+
+  std::promise<ResultCode> compilePromise;
+  size_t compilesLeft{numDevices};
+  for (int i = 0; i < numDevices; ++i) {
+    devices[i]->addNetwork(
+        &module, functionMap,
+        [i, &compilePromise, &compilesLeft](const Module *, ResultCode code) {
+          if (code != ResultCode::Ready) {
+            llvm::errs() << "Failed to compile model for device " << i << ".\n";
+            compilePromise.set_value(code);
+          }
+
+          llvm::outs() << "Successfully added to Device " << i << ".\n";
+          if (--compilesLeft == 0) {
+            compilePromise.set_value(ResultCode::Ready);
+          }
+        });
+  }
+
+  ResultCode result = compilePromise.get_future().get();
+  if (result != ResultCode::Ready) {
+    return 1;
+  }
+
+  std::error_code code;
+  llvm::sys::fs::directory_iterator dirIt(inputDirectory, code);
+  if (code.value()) {
+    llvm::errs() << "Couldn't read from directory: " << inputDirectory
+                 << " - code" << code.value() << "\n";
+    exit(code.value());
+  }
+
+  std::promise<void> finished;
+
+  size_t started = 0;
+  std::atomic<size_t> returned{0};
+
+  size_t max = 1000;
+
+  while (started++ < max) {
+    if (code.value() != 0 || dirIt == llvm::sys::fs::directory_iterator()) {
```

Comment:
`directory_iterator()` is the end iterator I'm assuming?

Reply:
Yeah directory_iterator doesn't have an end() it seems, it just uses the "empty iterator".